src/routing_table/mod.rs

```diff
@@ -257,6 +260,71 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
         }
     }
 
+    pub fn new_with_prefixes<U: IntoIterator<Item = Prefix<T>>>(our_name: T,
+                                                                min_group_size: usize,
+                                                                prefixes: U)
+                                                                -> Result<Self, Error> {
+        let mut groups = HashMap::new();
+        let mut our_group_prefix = Prefix::new(0, our_name);
+        for prefix in prefixes {
+            if prefix.matches(&our_name) && prefix.bit_count() > our_group_prefix.bit_count() {
+                our_group_prefix = prefix;
+            }
+            let _ = groups.insert(prefix, HashSet::new());
+        }
+        let result = RoutingTable {
+            our_name: our_name,
+            min_group_size: min_group_size,
+            our_group_prefix: our_group_prefix,
+            groups: groups,
+            needed: HashSet::new(),
+        };
+        if let Err(error) = result.satisfies_invariant() {
+            Err(Error::InvariantViolation(error))
+        } else {
+            Ok(result)
+        }
+    }
+
+    fn satisfies_invariant(&self) -> Result<(), &'static str> {
+        if !self.our_group_prefix.matches(&self.our_name) {
+            return Err("our prefix does not match our name");
```

Comment:
I defined `InvariantViolation` as `InvariantViolation(&str)`, so we can return the reason in the error - logging would have to duplicate each message (`warn!(message); return Error::IV(message);`). Or should I remove that and only log?